oak-doc/src/site/markdown/security/authorization/bestpractices.md

```diff
@@ -0,0 +1,316 @@
+         deny  readers    jcr:read restriction(rep:subtrees, /sensitive_info)        # what about editors or a subject being both reader and editor?
+         allow everyone   jcr:read restriction(rep:subtrees, /public, /also_public) # different public folders??
+         
+         # ... and what happens with a new node /content/public/abc/sensitive_info?
+      end 
+
+### Define Roles and Tasks
+
+Finally, write down basic characteristics and demands of your application without getting into access control details 
+or making any assumptions on how your needs will reflected in the repository:
+
+- what roles are present
+- what kind of tasks are those roles designed to perform
+- define if you have services accessing the repository and what kind of tasks they need to complete 
+
+Note, that this document should be human readable not go into implementation details:
+Instead of writing principal 'content-authors' needs jcr:write on /content, define that you have an asset 'content',
+define what kind of data it contains and how sensitive the data is (similar to the threat model).
+Then identify what roles are going to interact with this data and how they interact: for example you may identify 
+a role that just reading data, a second role that is expected to read and write and a third one that is will only 
+approve new content and publish it).
+
+## General Best Practices
+
+### Know how to get what you need
+
+Familiarize yourself with JCR access control management and Oak authorization design and extensions before starting 
+to edit the permission setup of your Oak installation. This will help you avoid common pitfalls. If you find yourself 
+granting your _content-writers_ role full access to just make it work, you probably left your application vulnerable.
+
+- JCR Specification sections [Access Control Management](https://s.apache.org/jcr-2.0-spec/16_Access_Control_Management.html) 
+and [Permissions and Capabilities](https://s.apache.org/jcr-2.0-spec/9_Permissions_and_Capabilities.html)
+- [Oak Authorization Documentation](../authorization.html) with separate sections for [Access Control Management](../accesscontrol.html) and [Permission Evaluation](../permission.html).
+- Exercises for authorization topics below https://github.com/apache/jackrabbit-oak/tree/trunk/oak-exercise/src/test/java/org/apache/jackrabbit/oak/exercise/security/authorization
+
+### Principle of least privilege
+
+Keep in mind that not having any permissions granted is equivalent to denying everything (which is in 
+this case redundant). Start without any access and then keep granting permissions as needed, following the 
+[principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege).
+In other words: only grant the minimal set of privileges required to perform a particular task.
+
+### Verification
+
+Write tests upfront and verify that for each role and task the expected effective permissions (see definition of roles) are 
```

Comment:
Any recommendation how such tests could look like? Examples?

Reply:
only pseudo-code.... but i can add that